File: python_sdk/examples/example_upload_reality_data.py
# Copyright (c) Bentley Systems, Incorporated. All rights reserved.
# See LICENSE.md in the project root for license terms and full copyright notice.

import os
import time
from dotenv import load_dotenv
from reality_capture.service.data_handler import RealityDataHandler
from reality_capture.service.service import RealityCaptureService, RealityDataCreate
from reality_capture.service.reality_data import Type
from utils import token_factory


def sleep(ms: int):
    time.sleep(ms / 1000)


def run_upload_example():
    """
    This example shows how to upload and download data from/to an iTwin using the Reality Capture SDK in Python.
    """

    # Inputs to provide. Please, adapt values
    images_path = "D:/Datasets/Helico/Images"
    output_path = "D:/Downloads/Helico/Images"
    images_name = "Reality Capture SDK context scene example"

    # Load environment variables
    load_dotenv()

    itwin_id = os.getenv("ITWIN_ID", "")
    client_id = os.getenv("CLIENT_ID", "")
    client_secret = os.getenv("CLIENT_SECRET", "")

    client_info = token_factory.ClientInfo(client_id=client_id, env="prod", secret=client_secret)
    token_factory_service = token_factory.ServiceTokenFactory(client_info)
    reality_data_handler = RealityDataHandler(token_factory_service)
    reality_capture_service_data = RealityCaptureService(token_factory_service)
    print("Reality Data handler initialized")

    try:
        print("Upload images in", itwin_id)
        reality_data_create = RealityDataCreate(iTwinId=itwin_id, type=Type.CC_IMAGE_COLLECTION,
                                                displayName=images_name)
        ret = reality_capture_service_data.create_reality_data(reality_data_create)
        if ret.is_error():
            print(f"Can't create reality data {images_name} in iTwin {itwin_id}")
            return
        reality_data_id = ret.value.id
        response = reality_data_handler.upload_data(reality_data_id, images_path, "", itwin_id)
        if response.is_error():
            print("Failed to upload reality data:", response.error.error.message)
            return
        print("Successfully uploaded images")

        print("Downloading images in", output_path)
        response_download = reality_data_handler.download_data(reality_data_id, output_path, "", itwin_id)
        if response_download.is_error():
            print("Failed to download reality data:", response_download.error.error.message)
            return
        print("Successfully downloaded images")
    except Exception as error:
        print(error)


if __name__ == "__main__":
    run_upload_example()

File: typescript/examples/src/example_upload_reality_data.ts
/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

import * as dotenv from "dotenv";
import * as fs from "fs";
import { ServiceAuthorizationClient } from "@itwin/service-authorization";
import { RealityCaptureService, RealityDataCreate, RealityDataHandler, Type } from "@itwin/reality-capture";

export async function sleep(ms: number) { return new Promise(resolve => setTimeout(resolve, ms)); }

async function runUploadExample() {
  /**
   * This example shows how to upload and download data from/to an iTwin
   */

  // Inputs to provide. Please, adapt values
  const imagesPath = "D:/Datasets/Helico/Images";
  const outputPath = "D:/Downloads/Helico/Images";

  const imagesName = "Reality Capture SDK context scene example";

  // Script
  dotenv.config();

  const iTwinId = process.env.IMJS_ITWIN_ID ?? "";
  const clientId = process.env.IMJS_CLIENT_ID ?? "";
  const secret = process.env.IMJS_CLIENT_SECRET ?? "";
  const issuerUrl = "https://ims.bentley.com";

  const authorizationClient = new ServiceAuthorizationClient({
    clientId: clientId,
    scope: "itwin-platform",
    clientSecret: secret,
    authority: issuerUrl,
  });

  const realityDataHandler = new RealityDataHandler(authorizationClient);
  console.log("Reality Data handler initialized");

  const realityCaptureService = new RealityCaptureService(authorizationClient);
  console.log("Reality Capture service initialized");

  try {
    if(!fs.existsSync(imagesPath)) {
      throw new Error(imagesPath + " does not exist");
    }

    console.log("Upload images in ", iTwinId);
    const realityDataCreate: RealityDataCreate = { iTwinId: iTwinId, displayName: imagesName, type: Type.CC_IMAGE_COLLECTION };
    const createdRealityDataResponse = await realityCaptureService.createRealityData(realityDataCreate);
    if (createdRealityDataResponse.isError()) {
      throw new Error("Failed to create reality data : " + createdRealityDataResponse.error!.error.message);
    }
    const createdRealityData = createdRealityDataResponse.value!;
    const response = await realityDataHandler.uploadData(createdRealityData.id, imagesPath, "", iTwinId);
    if (response.isError()) {
      throw new Error("Failed to upload reality data : " + response.error!.error.message);
    }
    console.log("Successfully uploaded images")

    console.log("Downloading images in ", outputPath);
    const responseDownload = await realityDataHandler.downloadData(createdRealityData.id, outputPath, "", iTwinId);
    if (responseDownload.isError()) {
      throw new Error("Failed to download reality data : " + responseDownload.error!.error.message);
    }
    console.log("Successfully downloaded images");
  }
  catch (error: any) {
    console.log(error);
  }
}

runUploadExample();